### DISCUSSION TITLE
Cannot access env variables

### DISCUSSION THREAD

**Discussion by samuelshumake** [vercel/next.js#72203] [Help] [answered]
### Summary

I'm trying to set up a vercel project with a postgres database, but I can't seem to access the env variables and actually access the database.

I'm in the root directory, and I run `vercel env pull .env.development.local` which makes my directory look like this:

```
webapp
-.env.development.local
-src/
--app/
---api/
----create-project/
-----route.ts
```

In my route.ts file, this is the code I have:

```
import { sql } from '@vercel/postgres';

export default async function createProject() {
    console.log({
        POSTGRES_URL: process.env.NEXT_PUBLIC_POSTGRES_URL,
        POSTGRES_URL_NON_POOLING: process.env.NEXT_PUBLIC_POSTGRES_URL_NON_POOLING
    });
}
```

But the console.log returns undefined for both POSTGRES_URL and POSTGRES_URL_NON_POOLING.

I'm very new to Vercel and NextJS, any ideas on how to fix this? To start local development, I run `vercel dev`

**Comment by icyJoseph** [comment]
Are the env vars present in the file pulled from Vercel?

**Reply by samuelshumake** [reply]
I believe so, in the *Settings -> Environment Variables* section, I see the list of them with hidden values. There's no "Shared Environment Variables" though.

**Reply by icyJoseph** [reply]
Are these available to the development environment as well? And do the contents of the `.env.development.local` show anything at all? (this is the most important bit)

**Reply by samuelshumake** [reply]
Yes, they're checked for the development environment. The .env.development.local file itself also has content in it as well, it's just when I try to get my program to print, process.env is empty

**Reply by icyJoseph** [reply]
Does it change anything if you do `npx next dev` instead?

**Reply by samuelshumake** [reply]
Unfortunately still returning undefined

**Reply by icyJoseph** [reply]
Whoa that's unexpected.

Could you post a screenshot of your project tree?

**Reply by samuelshumake** [reply]
I apologize for the messy setup, very new to NextJS and Vercel. I've tried putting the .env file at about every level
<img width="580" alt="Screenshot 2024-11-07 at 11 45 52 AM" src="https://github.com/user-attachments/assets/883c4b46-ef52-4959-b4cb-73fab1e45c89">

**Reply by icyJoseph** [reply]
I guess, there's some kind of conflict with the `.env.development.local` at the root of the project, and the `.env.development.local`, at the webapp project? If you cd into webapp, and run `npx next dev` does it work?

**Reply by samuelshumake** [reply]
`npx next dev` in the webapp directory doesn't have it working either unfortunately.

**Reply by samuelshumake** [reply, marked as answer]
@icyJoseph I figured this out! I was going about it the wrong way. I think I was attempting to call the server database function directly from the client component (if that makes sense). I restructured my folder setup and now on file submission, I have this:

```
const handleSubmit = async (file) => {
    // e.preventDefault();

    try {
      const response = await fetch('/api/upload', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ content: file }), // Send file content to server
      });

      if (response.ok) {
        console.log(response);
      } else {
        console.error('File upload failed');
      }
    } catch (error) {
      console.error('An error occurred:', error);
    }
  };
  ```
  
  Then, in an `api/upload/route.ts` file, I have this:
  
  ```
  import { sql } from '@vercel/postgres';

export async function POST(request: Request) {
  const res = await request.json()

  const data = await sql`INSERT INTO test (name) VALUES ('hello world')`;

  return Response.json({ res })
}
```

And that seems to be pushing to my database correctly!

**Reply by icyJoseph** [reply]
Ohh yah, you'd need to prefix the env vars with NEXT_PUBLIC_ to have them available on the client, but you probably didn't want that either. Good job!
